src/runtime/containerd-shim-v2/create.go

```diff
@@ -259,3 +272,80 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
 	}
 	return nil
 }
+
+func configureNonRootHypervisor(runtimeConfig *oci.RuntimeConfig) error {
+	userName, err := createVmmUser()
+	if err != nil {
+		return err
+	}
+
+	u, err := user.Lookup(userName)
+	if err != nil {
+	    return err
+    }
+
+    uid, err := strconv.Atoi(u.Uid)
+    if err != nil {
+        return err
+    }
+    gid, err := strconv.Atoi(u.Gid)
+    if err != nil {
+        return err
+    }
+	runtimeConfig.HypervisorConfig.Uid = uint32(uid)
+	runtimeConfig.HypervisorConfig.Gid = uint32(gid)
+
+	rootDir := path.Join("/run/user/", fmt.Sprint(uid))
+	dir, err := os.Stat(rootDir)
+	if os.IsNotExist(err) {
+	    if err = os.Mkdir(rootDir, vc.DirMode); err != nil {
+	        return err
+        }
+		if err = syscall.Chown(rootDir, uid, gid); err != nil {
+			return err
+		}
+    }
+    if dir != nil && !dir.IsDir() {
+    	return fmt.Errorf("%s is expected to be a directory", rootDir)
+	}
+
+	if err := os.Setenv("XDG_RUNTIME_DIR", rootDir); err != nil {
+		return err
+	}
+
+	info, _ := os.Stat("/dev/kvm")
+	var Gid uint32
+	if stat, ok := info.Sys().(*syscall.Stat_t); ok {
+		Gid = stat.Gid
+	} else {
+		return fmt.Errorf("failed to get the gid of dev/kvm")
+	}
+	runtimeConfig.HypervisorConfig.Groups = append(runtimeConfig.HypervisorConfig.Groups, Gid)
+
+	return nil
+}
+
+func createVmmUser() (string, error) {
+    var (
+    	err error
+        userName string
+    )
+    maxAttempt := 5
+    attempt := 0
+
+    for ok := true; ok; ok = err != nil && attempt < maxAttempt {
+        userName = fmt.Sprintf("qemu-%v", rand.Intn(100000))
+        _, err = utils.RunCommand([]string{"useradd", userName})
```

Comment:
I consider that this is prone to hacks if the PATH environment variable is not configured correctly, for security reasons I'd prefer to use the full path `/usr/sbin/useradd`